### PR TITLE
NET-84: Trusted MQTT plugin

### PR DESCRIPTION
**Pull request by teogeb**
New MQTT plugin which uses a trusted `StreamrClient`. 

Added also two helper utilities for plugins:

- `streamrClient`:  loops back to the same Broker instance via WebSocket plugin (to prototype Brubeck-style plugins)
- `apiAuthenticator`: configuration of valid apiKeys which plugins can use to authenticate requests

Renamed the deprecated MQTT plugin to `legacyMqtt`.

**Breaking changes** in config files: `apiAuthentication` added , `mqtt` plugin renamed

### CODE REVIEW

File: packages/broker/src/plugins/mqtt/MqttServer.ts [hunk 1]
```diff
@@ -1,303 +1,93 @@
...
-        })
-
-        connection.on('disconnect', () => {
-            logger.debug('client disconnected')
-        })
-
-        connection.on('publish', (publishPacket) => {
-            this.handlePublishRequest(connection, publishPacket)
+        this.aedes.on('publish', (packet: IPublishPacket, client: aedes.Client) => {
+            if (client !== null) {  // is null if the this server sent the message
```
Comment (harbu) [inline packages/broker/src/plugins/mqtt/MqttServer.ts:33]: :+1:

File: packages/broker/src/plugins/mqtt/MqttServer.ts [hunk 2]
```diff
@@ -1,303 +1,93 @@
...
-            connection.markAsDead()
-            this._closeConnection(connection)
+export class MqttServer {
+    private static COMMAND_PUBLISH = 'publish'
+    private static QOS_EXACTLY_ONCE = 2
+    private static BAD_USERNAME_OR_PASSWORD = 4
+    private static NOT_AUTHORIZED = 5
+
+    private port: number
+    private aedes: aedes.Aedes
```
Comment (harbu) [inline packages/broker/src/plugins/mqtt/MqttServer.ts:23]:
```suggestion
    private readonly port: number
    private readonly aedes: aedes.Aedes
```

File: packages/broker/src/plugins/mqtt/MqttServer.ts [hunk 3]
```diff
@@ -1,303 +1,93 @@
...
-        const stream = this.streams.get(streamId, 0)
-
-        if (stream) {
-            const object = {
-                cmd: 'publish',
-                topic: stream.name,
-                payload: JSON.stringify(streamMessage.getParsedContent())
+    private static createAuthenicationHandler(apiAuthenticator: ApiAuthenticator): aedes.AuthenticateHandler {
+        return (_client: aedes.Client, _username: Readonly<string>|undefined, password: Readonly<Buffer>|undefined, done: (error: aedes.AuthenticateError|null, success: boolean|null) => void) => {
+            if (apiAuthenticator!.isValidAuthentication(password?.toString())) {
```
Comment (harbu) [inline packages/broker/src/plugins/mqtt/MqttServer.ts:83]: Would it work without exclamation mark?
```suggestion
            if (apiAuthenticator.isValidAuthentication(password?.toString())) {
```

File: packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            },
+            apiAuthentication: {
+                keys: [MOCK_API_KEY]
+            }
+        })
+    })
+
+    afterAll(async () => {
+        await broker.close()
+        await tracker.stop()
```
Comment (harbu) [inline packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts:61]: With `allSettled` you can make sure that the if e.g. `await broker.close()` would throw then `tracker.stop()` also gets called.
```suggestion
        await Promise.allSettled([
            broker.close(),
            tracker.stop()
        ])
```

File: packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        mqttClient = await createMqttClient()
+        topic = 'topic-' + Date.now()
+        await streamrClient.createStream({
+            id: getStreamId()
+        })
+    })
+
+    afterEach(async () => {
+        await mqttClient.end(true)
+        await streamrClient.ensureDisconnected()
```
Comment (harbu) [inline packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts:77]:
```suggestion
        await Promise.allSettled([
            mqttClient.end(true),
            streamrClient.ensureDisconnected()
        ])
```

File: packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            stream: getStreamId()
+        }, (message: any) => {
+            receivedMessage = message
+        })
+        mqttClient.publish(topic, createPayload())
+        await waitForCondition(() => receivedMessage !== undefined)
+        expect(receivedMessage.foo).toBe('from-client')
+    })
+
+    test('subscribe on MQTT client', async () => {
```
Comment (harbu) [inline packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts:100]: Nice tests :+1:

File: packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        let receivedTopic: string|undefined
+        let receivedJson: any 
+        mqttClient.on('message', (topic: string, payload: Buffer) => {
+            receivedTopic = topic
+            receivedJson = JSON.parse(payload.toString())
+        })
+        await mqttClient.subscribe(topic)
+        await wait(1000)
```
Comment (harbu) [inline packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts:108]: You sure the `wait` is needed here? I was wondering if `await mqttClient.subscribe(..)` above has some sort of ACK mechanism?

**Review by harbu — APPROVED** [review]
Nice stuff!

File: packages/broker/src/broker.ts [hunk 8]
```diff
@@ -164,7 +170,10 @@ export const startBroker = async (config: Config): Promise<Broker> => {
                 await stopServer(httpServer)
             }
             await Promise.all(plugins.map((plugin) => plugin.stop()))
-            return Promise.all([networkNode.stop(), volumeLogger.close()])
+            if (localStreamrClient !== undefined) {
+                await localStreamrClient!.ensureDisconnected()
```
[the remark this answers is not in the document]
Comment (teogeb) [reply packages/broker/src/broker.ts:174]: True, fixed

File: packages/broker/src/plugins/mqtt/MqttPlugin.ts [hunk 9]
```diff
@@ -1,46 +1,30 @@
...
-            new net.Server().listen(this.pluginConfig.port).on('listening', () => logger.info(`Mqtt plugin listening on ${this.pluginConfig.port}`)),
-            this.pluginConfig.streamsTimeout,
-            this.networkNode,
-            new StreamFetcher(this.brokerConfig.streamrUrl),
-            this.publisher,
-            this.metricsContext,
-            this.subscriptionManager
-        )
+        this.server = new MqttServer(this.pluginConfig.port, this.apiAuthenticator)
+        const bridge = new Bridge(this.streamrClient!, this.server, this.pluginConfig.streamIdDomain ?? undefined)
```
[the remark this answers is not in the document]
Comment (teogeb) [reply packages/broker/src/plugins/mqtt/MqttPlugin.ts:21]: Added

File: packages/broker/src/plugins/mqtt/Bridge.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    streamIdDomain?: string
+
+    constructor(streamrClient: StreamrClient, mqttServer: MqttServer, streamIdDomain?: string) {
+        this.streamrClient = streamrClient
+        this.mqttServer = mqttServer
+        this.streamIdDomain = streamIdDomain
+    }
+
+    onMessageReceived(topic: string, payload: string) {
+        const { message, metadata } = JSON.parse(payload)
```
[the remark this answers is not in the document]
Comment (teogeb) [reply packages/broker/src/plugins/mqtt/Bridge.ts:20]: Added error logging. More improvements here: https://linear.app/streamr/issue/NET-330/mqtt-plugin-error-handling-for-publish

File: packages/broker/src/plugins/mqtt/Bridge.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        this.streamrClient!.publish(this.getStreamId(topic), message, metadata?.timestamp)
```
[the remark this answers is not in the document]
Comment (teogeb) [reply packages/broker/src/plugins/mqtt/Bridge.ts:21]: Fixed

File: packages/broker/src/plugins/mqtt/Bridge.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    
+    onSubscribed(topic: string) {
+        logger.info('Client subscribed: ' + topic)
+        this.streamrClient!.subscribe(this.getStreamId(topic), (message: any, metadata: any) => {
```
[the remark this answers is not in the document]
Comment (teogeb) [reply packages/broker/src/plugins/mqtt/Bridge.ts:26]: Fixed

File: packages/broker/src/plugins/mqtt/MqttServer.ts (hunk 2 above)
Comment (teogeb) [reply packages/broker/src/plugins/mqtt/MqttServer.ts:23]: Fixed

File: packages/broker/src/plugins/mqtt/MqttServer.ts (hunk 3 above)
Comment (teogeb) [reply packages/broker/src/plugins/mqtt/MqttServer.ts:83]: Fixed

File: packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts (hunk 4 above)
Comment (teogeb) [reply packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts:61]: Fixed

File: packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts (hunk 5 above)
Comment (teogeb) [reply packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts:77]: Fixed

File: packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts (hunk 7 above)
Comment (teogeb) [reply packages/broker/test/integration/plugins/mqtt/MqttPlugin.test.ts:108]: Removed

File: packages/broker/src/plugins/legacyMqtt/MqttPlugin.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+import PLUGIN_CONFIG_SCHEMA from './config.schema.json'
+
+const logger = new Logger(module)
+
+export interface MqttPluginConfig {
+    port: number
+    streamsTimeout: number|null
+}
+
+export class MqttPlugin extends Plugin<MqttPluginConfig> {
```
[the remark this answers is not in the document]
Comment (teogeb) [reply packages/broker/src/plugins/legacyMqtt/MqttPlugin.ts:16]: If we can't remove the legacy plugins soon.

**Pull request merged**